tiledb/api/proxy/proxy.h

```diff
@@ -0,0 +1,917 @@
+ * function `shutdown()`
+ */
+template <bool has_shutdown>
+class ProxyStateMachine {
+ protected:
+  using lock_type = std::unique_lock<std::mutex>;
+
+ private:
+  /**
+   * The current state of the state machine.
+   */
+  ProxyState state_;
+
+  /**
+   * The number of currently outstanding accessors.
+   *
+   * @invariant n_accessors > 0 if and only if the state is `access` or
+   * `last_access`
+   */
+  unsigned int n_accessors{0};
+
+  /**
+   * Mutex to synchronize queue operations.
+   *
+   * This is the first of two mutexes. Whenever both mutexes must be locked,
+   * this one must be locked first.
+   */
+  std::mutex m_queue_{};
+
+ protected:
+  /**
+   * Mutex to synchronize state transitions.
+   *
+   * The is the second of two mutexes. Whenever both mutexes must be locked,
+   * this one must locked second.
+   */
+  std::mutex m_state_{};
+
+ private:
+  /**
+   * The internal event queue.
+   */
+  std::queue<ProxyEvent> queue_{};
+
+  /*
+   * Event processing states
+   *
+   * In addition to the state machine proper, event processing has an implicit
+   * state machine that depends upon the lock states of the two mutexes. There
+   * are four states; each state has two possible outgoing transitions depending
+   * on which mutex is changing. Thus there are eight possible transitions but
+   * only five are used. This state machine is embodied in the control flow of
+   * `event()`.
+   *
+   * ```
+   *                            Queue
+   *                    Unlocked     Locked
+   *                +------------------------------
+   *                |           <--
+   *       Unlocked |    Quiet  --> Receive
+   * State          |      ^           |
+   *                |      |           v
+   *         Locked |   Change <-- Hand-off
+   *                |
+   * ```
+   *
+   * Event processing does not have a separate thread. All events are processed
+   * by the thread that calls `event()`. A call to `event` may result a few
+   * kinds of behavior.
+   *
+   * 1. (Ordinary) The queue is empty and stays empty while the state machine
+   *    processes the event. This results in one single cycle.
+   * 2. The queue is not empty and the event is non-blocking. The message is
+   *    enqueued. This event moves from `Receive` to `Quiet` and returns.
+   * 3. The queue is not empty and the event is blocking. The message is
+   *    enqueued and blocks waiting to move to 'Hand-off'. The loop continues
+   *    until the queue is empty.
+   *
+   * Put more simply, the `event()` calls for blocking events are able to handle
+   * queued non-blocking ones.
+   *
+   * This behavior is necessary to support prompt shutdown. The `shutdown` event
+   * is non-blocking so that a top-level shutdown thread is not at risk of
+   * blocking on any potentially long-lived state transitions. The `construct`
+   * event is considered long lived because it may involve storage I/O
+   * operations.
+   */
+
```

Comment:
Does a blocking event need to be processed in order for all non-blocking events to be processed?  I.e., is the queue guaranteed to be emptied?  This seems to imply that a blocking event is needed at some point to flush the non-blocking events.